content/post/postgres/rfc-pgxn-metadata-sketch.md

```diff
@@ -0,0 +1,959 @@
+    "registry": "pgxman.com",
+    "type": "apt",
+    "url": "https://pgxman.com/x/pair",
+    "command": "pgxman install pair"
+  },
+  {
+    "registry": "database.dev",
+    "type": "sql",
+    "url": "https://database.dev/example/pair",
+    "command": "SELECT dbdev.install('example-pair');"
+  },
+]
+```
+
+Other Options:
+
+*   Information about installing from a registry that doesn't offer URLs, like
+    Homebrew, where one just needs to know what the formula is called. Maybe
+    the same should apply to Apt?
+
+### Metrics and Reports
+
+This object includes data such as number of downloads, build and test status
+on various Postgres/OS/version combinations, binary packaging distributions,
+test coverage, security certification, vulnerability detection, quality
+metrics and user ratings, and more. Some of these bits of data might include
+badges, e.g., for ratings or security assessments.
+
+In the broader ecosystem, it would be the responsibility of the root registry
+to ensure that data comes only from trusted sources. However this data will be
+publicly readable, allowing *any* downstream applications to extend and
+publish metrics and reports with their own information.
+
+Example:
+
+``` json
+{
+    "aggregates": {
+      "downloads": 32
+    },
+    "sources": {
+      "smoke-tester.example.com": {
+        "stats": {
+          "passes": 32,
+          "fails": 8,
+          "unknown": 0
+        },
+        "link": "https://smoke-tester.example.com/extension/pair",
+        "badge": "https://smoke-tester.example.com/extension/pair.svg"
+      },
+      "pgt.dev": {
+        "downloads": 12,
+        "tags": ["two", "kv"],
+        "categories": ["Analytics"]
+      },
+      "pgxman.com": {
+        "downloads": 20
+      },
+      "ratings.example.com": {
+        "stags": {
+          "ratings": 324,
+          "average": 3.2,
+          "comments": 13
+        },
+        "link": "https://ratings.example.com/extension/pair",
+        "badge": "https://ratings.example.com/extension/pair.svg"
+      },
+      "stats.example.com": {
+        "contributors": 2,
+        "stars": 14,
+        "last_active": "2024-01-12",
+      },
+      "security-scan.example.com": {
+        "advisories": {
+          "high": 2,
+          "moderate": 11,
+          "low": 1,
+          "cves": 0
+        },
+        "link": "https://security-scan.example.com/extension/pair",
+        "badge": "https://security-scan.example.com/extension/pair.svg"
+      }
+    }
+}
+```
+
+**Notes:**
+
+*   The `aggregates` section aggregates results from multiple sources, for
+    example summing all downloads or averaging ratings. The list of items to
+    aggregate could evolve regularly.
+
+*   Each key in `sources` identifies a trusted downstream source of
+    information. Each would have its own schema describing its objects and
+    their meaning, along with URI templates to link to. For example,
+    `stats.example.com` might have these templates:
```

Comment:
Sorry, I don't really follow, can you explain what "summation loops" means?

I'd like to come up with a federation model, but admit I wasn't really thinking about it here.

Reply:
Oooh. I hadn't even thought about the aggregation being bidirectional, but of course people would want that, if things are successful. Not an immediate goal, I think, because you're right, the technical infrastructure to prevent these kinds of loops will need some careful thought.